### PR TITLE
Modify `FusionDefinition.add_output` API to an include an `alias_input` option.

### PR DESCRIPTION
**Pull request by kevinstephano**
Enable Python Frontend to allow Outputs to alias Inputs in order to implement the running stats in operations like Batch Norm and Instance Norm.

Example Usage:

```
import torch
from nvfuser._C import FusionDefinition, Fusion, DataType

def fusion(fd : FusionDefinition) :
    t0 = fd.define_tensor(2)
    t1 = fd.ops.relu(t0)
    fd.add_output(t1, alias_input=t0)

fs = Fusion()
with FusionDefinition(fs) as fd:
    fusion(fd)

inputs = [
    torch.randn(128, 128, device='cuda'),
]

for _ in range(5):
    out = fs.execute(inputs)
```

### CODE REVIEW

**Comment by kevinstephano** [comment]
cc @jacobhinkle

**Comment by IvanYashchuk** [comment]
In BatchNorm inputs are updated in-place but the running stats are not returned as outputs. How to deal with this?
I attempted to implement it with a new [`ops.copy_to` function](https://github.com/pytorch/pytorch/pull/84545/files#diff-a4555e6dabf5ae1ce8fcd06f4cb33131dee0e151c8db012a611be5e747435948R1321) which underneath uses `aliasOutputToInput` and `addOutput`. This adds redundant outputs to the `execute` function, which are stripped away by counting the number of `ops.copy_to` calls.

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp [hunk 1]
```diff
@@ -140,18 +140,32 @@ void initNvFuserPythonBindings(PyObject* module) {
...
-                {self.recordingState(output())}));
-          })
+            if (alias_input.has_value()) {
+              self.defineRecord(new nvfuser::OutputRecord<Nvf::Val>(
+                  {self.recordingState(output()), self.recordingState(alias_input.value()())}));
+            } else {
+              self.defineRecord(new nvfuser::OutputRecord<Nvf::Val>(
+                  {self.recordingState(output())}));
+            }
+          },
```
Comment (jacobhinkle) [inline third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:147]: Might be nice to add a docstring here by inserting a line like
```cpp
"Declare scalar argment 'output' as an output. If given, 'alias_input' will be aliased by 'output', representing in-place modification of 'alias_input'.",
```

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp [hunk 2]
```diff
@@ -140,18 +140,32 @@ void initNvFuserPythonBindings(PyObject* module) {
           })
       .def(
           "add_output",
-          [](nvfuser::FusionDefinition& self, nvfuser::Scalar output) {
+          [](nvfuser::FusionDefinition& self, nvfuser::Scalar output, c10::optional<nvfuser::Scalar> alias_input = c10::nullopt) {
```
Comment (jacobhinkle) [inline third_party/nvfuser/csrc/python_frontend/python_bindings.cpp]: Scalars can be aliased in nvfuser, but what is the meaning of that in Python where scalars are immutable? Maybe I am misunderstanding what this represents...

**Review by jacobhinkle — APPROVED** [review]

File: test/test_nvfuser_frontend.py [hunk 3]
```diff
@@ -581,5 +583,25 @@ def fusion_func(fd : FusionDefinition) :
...
+            t1 = fd.ops.add(t0, s0)
+            t2 = fd.ops.add(t1, s0)
+            fd.add_output(t1, alias_input=t0)
+            fd.add_output(t2)
+
+        inputs = [
+            torch.ones(4, 4, device='cuda'),
+        ]
+
+        nvf_out, _ = self.exec_nvfuser(fusion_func, inputs)
```
Comment (jjsjann123) [inline test/test_nvfuser_frontend.py:605]: I agree with Ivan that there's a need to hide the aliased output in order to match the semantics. Let's put an example for that as well.

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp [hunk 4]
```diff
@@ -140,18 +140,32 @@ void initNvFuserPythonBindings(PyObject* module) {
...
             FUSER_PERF_SCOPE("FusionDefinition.add_output (scalar)");
-            self.defineRecord(new nvfuser::OutputRecord<Nvf::Val>(
-                {self.recordingState(output())}));
-          })
+            if (alias_input.has_value()) {
+              self.defineRecord(new nvfuser::OutputRecord<Nvf::Val>(
+                  {self.recordingState(output()), self.recordingState(alias_input.value()())}));
+            } else {
+              self.defineRecord(new nvfuser::OutputRecord<Nvf::Val>(
+                  {self.recordingState(output())}));
```
Comment (jjsjann123) [inline third_party/nvfuser/csrc/python_frontend/python_bindings.cpp]: nitpick: instead of variadic record, maybe we can just have default for alias entries? Something like a `None`.

Maybe it's too late already, I'm not totally sure if there's already pattern in the code base using this :)

File: test/test_nvfuser_frontend.py (hunk 3 above)
Comment (jjsjann123) [reply test/test_nvfuser_frontend.py:605]: One more thing regarding returning a reference to aliased output.

IMHO, we should try to restrict the usage of aliases. Because we don't have proper alias analysis, so it's totally not a safe thing to use and we should make it CRYSTAL CLEAR in our APIs.
Let's just NOT support returning a reference to the aliased input. So it's less likely that a general case aliased update is accidentally implemented with nvfuser python frontend.

I know it's a weak argument... We can also leave the code as-is and start a separate discussion on how to better communicate it with users.

**Review by jjsjann123 — APPROVED** [review]
Overall looks good to me.

Have some minor comments and opinions on how the API and usage example could be revisited.

**Comment by jjsjann123** [comment]
> In BatchNorm inputs are updated in-place but the running stats are not returned as outputs. How to deal with this? I attempted to implement it with a new [`ops.copy_to` function](https://github.com/pytorch/pytorch/pull/84545/files#diff-a4555e6dabf5ae1ce8fcd06f4cb33131dee0e151c8db012a611be5e747435948R1321) which underneath uses `aliasOutputToInput` and `addOutput`. This adds redundant outputs to the `execute` function, which are stripped away by counting the number of `ops.copy_to` calls.

I think we wouldn't want it to be done at user level. That gives the impression that aliases are well supported in codegen, which it is NOT!!!
This is a hack and we should make it more obvious if we can. That's why I was arguing that we should change the example and exposure of the API, so user, aka op implementer, don't have to tinker with aliases trying to match the schema of the operator.

I'm getting very scared whenever aliases are used.

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp (hunk 1 above)
Comment (jacobhinkle) [reply third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:147]: This could be a place to put a big scary warning.

Comment (kevinstephano) [reply third_party/nvfuser/csrc/python_frontend/python_bindings.cpp:147]: I got rid of the Scalar aliasing.  That did not make sense to me after thinking about it.

File: third_party/nvfuser/csrc/python_frontend/python_bindings.cpp (hunk 4 above)
Comment (kevinstephano) [reply third_party/nvfuser/csrc/python_frontend/python_bindings.cpp]: I got rid of the Scalar aliasing.
